### PR TITLE
Favorite System v1.0

### PR DESCRIPTION
**Pull request by cafeoh**
#### Summary
```SUMMARY: Features "Adds a fully functional item favoriting system"```

_Back with more better features and 70% cleaner code!_

gif
gif

(Thanks to @anothersimulacrum for his save by the way! ❤️ )
#### Purpose of change
Allows to favorite items from every menu including the advanced item menu (but not from the Compare and Multidrop menus). In addition to selecting a single item via the inventory menu (`Enter` -> `f`) there's now a new default keybinding to `*` that favorites the full stack under the cursor.

Two menu behaviours were added to make this actually useful (beyond just the visual aspect) :
- Category selection in the drop menu, on successive selection, first selects all non favorite item, then all favorite items, then deselects. This allows you, for example, after returning from a profitable food hoarding run to drop everything except your jerky and V8.
- Advanced inventory works similarly when using the drop all (`,`) key from your inventory, where it only first empties out all non favorite item on the first press, and then drop all the rest of your favourite gear (with a warning).

#### Describe the solution
Changed item stacking, item selector, and other UI stuff.

#### Future additions
- Implement the `,` key in the multidrop menu (might be redundant with the advanced inventory, but it'd be pretty straightforward if people are interested 🤷‍♂️ )
- When dropping items out of necessity (dropping your faithful duffel bag for example), make non favorite items drop first. Haven't looked into it, should be pretty easy though.
- Add favoriting from drop menu (To be honest that'd be a pain for a few reasons, as favoriting items can cause stacks to merge and requires to reload all items for safety, therefore requiring loops outside input handling, partial refresh/caching items and just a lot of nastiness)
- Find a better word for "favoriting"? I don't know it's starting to lose all its meaning to me. Favoriting. Favoriting, favoriting. Yeah I hate it. 😡

### CODE REVIEW

**Comment by Wokko1** [comment]
Has any other PR received that many upvotes in so little time?

**Review by mlangsdorf — COMMENTED** [review]
my knowledge of the inventory system is just enough to get me in trouble, but nothing in this looks obviously bad.

**Review by ifreund — COMMENTED** [review]
Code looks great to me, haven't compiled or tested though.  Nice work!

File: src/inventory.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@
 #include "vpart_reference.h"
 
 const invlet_wrapper
-inv_chars( "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"#&()*+.:;=@[\\]^_{|}" );
+inv_chars( "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"#&()+.:;=@[\\]^_{|}" );
```
[the remark this answers is not in the document]
Comment (cafeoh) [reply src/inventory.cpp]: I tested it, and basically the items just keep the invlet. As usual, invlets take precedence over key bindings, so it means you won't be able to use the `*` bind until you unassign the item, but I don't think it's a big deal. I thought about cleaning the invlets manually (making sure they match), but I'm not convinced it'd really be worth it to have that piece of laying around.

I wanted to improve handling of invlet attribution later so it's not as hardcoded (there's no verification at the moment), but that's for another PR.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/latest-experimental-features/5582/1106

**Pull request merged**